lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
-]]
-local function set(rbx, key, value)
-	rbx[key] = value
+	tree.tasksRunning = false
 end
 
---[[
-	Sets a property on a Roblox object, following Roact's rules for special
-	case properties.
-
-	This function can throw a couple different errors. In the future, calls to
-	_setRbxProp should be wrapped in a pcall to give better errors to the user.
-
-	For that to be useful, we'll need to attach a 'source' property on every
-	element, created using debug.traceback(), that points to where the element
-	was created.
-]]
-function Reconciler._setRbxProp(rbx, key, value, element)
-	if type(key) == "string" then
-		-- Regular property
+local function mountTree(element, parentRbx, key)
+	assert(Type.is(element, Type.Element))
+	assert(typeof(parentRbx) == "Instance" or typeof(parentRbx) == "nil")
+
+	local tree = {
+		[Type] = Type.Tree,
+		tasks = {},
+		taskIndex = 1,
+		tasksRunning = false,
+		connections = {},
+		mounted = true,
+		rootNode = nil,
+	}
+
+	if ASYNC_SCHEDULER then
+		tree.renderStepId = getGuid()
+		RunService:BindToRenderStep(tree.renderStepId, Enum.RenderPriority.Last.Value, function()
+			processTreeTasksAsync(tree, ASYNC_BUDGET_PER_FRAME)
+		end)
+	end
 
-		local success, err = pcall(set, rbx, key, value)
+	scheduleTask(tree, taskMountNode({
+		element = element,
+		key = key,
+		parentNode = nil,
+		parentRbx = parentRbx,
+		isTreeRoot = true,
+	}))
 
-		if not success then
-			local source = element.source or DEFAULT_SOURCE
+	return tree
+end
 
-			local message = ("Failed to set property %s on primitive instance of class %s\n%s\n%s"):format(
-				key,
-				rbx.ClassName,
-				err,
-				source
-			)
+local function unmountTree(tree)
+	assert(Type.is(tree, Type.Tree))
+	assert(tree.mounted, "not mounted")
 
-			error(message, 0)
-		end
-	elseif type(key) == "table" then
-		-- Special property with extra data attached.
+	tree.mounted = false
 
-		if key.type == Event then
-			Reconciler._singleEventManager:connect(rbx, key.name, value)
-		elseif key.type == Change then
-			Reconciler._singleEventManager:connectProperty(rbx, key.name, value)
-		else
-			local source = element.source or DEFAULT_SOURCE
+	if ASYNC_SCHEDULER then
+		RunService:UnbindFromRenderStep(tree.renderStepId)
+	end
 
-			-- luacheck: ignore 6
-			local message = ("Failed to set special property on primitive instance of class %s\nInvalid special property type %q\n%s"):format(
-				rbx.ClassName,
-				tostring(key.type),
-				source
-			)
+	-- TODO: Flush/cancel existing tasks and unmount asynchronously?
 
-			error(message, 0)
-		end
-	elseif type(key) ~= "userdata" then
-		-- Userdata values are special markers, usually created by Symbol
-		-- They have no data attached other than being unique keys
+	scheduleTask(tree, taskUnmountNode({
+		node = tree.rootNode,
+	}))
 
-		local source = element.source or DEFAULT_SOURCE
+	-- For now, flush the entire tree and unmount synchronously
+	processTreeTasksSync(tree)
```

Comment:
Not that this is too much of a problem, but the previous call to scheduleTask may have already called processTreeTasksSync.